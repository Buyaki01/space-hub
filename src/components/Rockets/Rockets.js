import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import './Rockets.scss';
import { fetchRockets } from '../../redux/rockets/rocketReducer';
import { reserveRockets } from '../../redux/rockets/rocketsAction';

const Rockets = () => {
  const dispatch = useDispatch();
  const rockets = useSelector(({ rocketsReducer }) => rocketsReducer.rockets);

  useEffect(() => {
    if (rockets.length === 0) dispatch(fetchRockets());
  }, []);

  const handleClick = (id) => {
    dispatch(reserveRockets(id));
  };

  return (
    <div className="rockets-container">
      {rockets.map((rocket) => (
        <li key={rocket.id} className="rocket_list">
          <div><img src={rocket.flickr_images[0]} alt="rocket" /></div>
          <div>
            <h2>{rocket.rocket_name}</h2>
            <p>
              {rocket.reserved && <span>Reserved</span>}
              {rocket.description}
              <a href={rocket.wikipedia} target="_blank" rel="noreferrer noopener">Read More...</a>
            </p>
            {!rocket.reserved && <button type="button" onClick={() => handleClick(rocket.id)}>Reserve Rocket</button>}
            {rocket.reserved && <button type="button" className="cancel" onClick={() => handleClick(rocket.id)}>Cancel Reservation</button>}
          </div>
        </li>
      ))}
    </div>
  );
};

export default Rockets;
