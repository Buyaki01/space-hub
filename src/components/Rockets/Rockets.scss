h2 {
  font-size: 1.5rem;
  color: green;
}
