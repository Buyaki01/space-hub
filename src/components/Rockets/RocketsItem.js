import PropTypes from 'prop-types';

const RocketsItem = ({
  id, image, name, reserved, desc, wiki, handleClick,
}) => (
  <>
    <div><img src={image} alt="rocket" /></div>
    <div>
      <h2>{name}</h2>
      <p>
        {reserved && <span>Reserved</span>}
        {desc}
        <a href={wiki} target="_blank" rel="noreferrer noopener">Read More...</a>
      </p>
      {!reserved && <button type="button" onClick={() => handleClick(id)}>Reserve Rocket</button>}
      {reserved && <button type="button" className="cancel" onClick={() => handleClick(id)}>Cancel Reservation</button>}
    </div>
  </>
);

RocketsItem.defaultProps = {
  handleClick() {},
  wiki: '',
  desc: '',
  name: '',
  image: '',
  reserved: '',
  id: 0,
};

RocketsItem.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  image: PropTypes.string,
  reserved: PropTypes.string,
  desc: PropTypes.string,
  wiki: PropTypes.string,
  handleClick: PropTypes.func,
};

export default RocketsItem;
