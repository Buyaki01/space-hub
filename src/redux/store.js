import { createStore, combineReducers, applyMiddleware } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import dragonsReducer from './dragons/dragonReducer';
<<<<<<< HEAD
import missionsReducer from './missions/missionsReducer';

const reducers = combineReducers({
  dragonsReducer,
  missionsReducer,
=======
import rocketsReducer from './rockets/rocketReducer';

const reducers = combineReducers({
  dragonsReducer,
  rocketsReducer,
>>>>>>> f8c89723c1210ea06b81e9f644d4ef10824c1ffd
});

const store = createStore(
  reducers,
  applyMiddleware(logger, thunk),
);

export default store;
