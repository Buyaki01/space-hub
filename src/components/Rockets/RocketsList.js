import { useSelector, useDispatch } from 'react-redux';

import './Rockets.scss';
import { reserveRockets } from '../../redux/rockets/rocketsAction';
import RocketsItem from './RocketsItem';

const Rockets = () => {
  const dispatch = useDispatch();
  const rockets = useSelector(({ rocketsReducer }) => rocketsReducer.rockets);

  const handleClick = (id) => {
    dispatch(reserveRockets(id));
  };

  return (
    <div className="rockets-container">
      {rockets.map((rocket) => (
        <li key={rocket.id} className="rocket_list" data-testid="rocket">
          <RocketsItem
            id={rocket.id}
            image={rocket.flickr_images[0]}
            name={rocket.rocket_name}
            reserved={rocket.reserved}
            desc={rocket.description}
            wiki={rocket.wikipedia}
            handleClick={handleClick}
          />
        </li>
      ))}
    </div>
  );
};

export default Rockets;
