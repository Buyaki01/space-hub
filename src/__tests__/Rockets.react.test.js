import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';

import Rockets from '../components/Rockets/RocketsList';
import RocketsItem from '../components/Rockets/RocketsItem';
import store from '../redux/store';

describe('Rockets', () => {
  test('should render component correctly', () => {
    const { asFragment } = render(<Provider store={store}><Rockets /></Provider>);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should conditionally render elements when buttons are clicked', () => {
    const { asFragment } = render(<Provider store={store}><RocketsItem /></Provider>);
    expect(asFragment()).toMatchSnapshot();
  });
});
