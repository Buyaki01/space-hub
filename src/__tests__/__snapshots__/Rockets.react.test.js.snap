// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets should conditionally render elements when buttons are clicked 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="rocket"
      src=""
    />
  </div>
  <div>
    <h2 />
    <p>
      <a
        href=""
        rel="noreferrer noopener"
        target="_blank"
      >
        Read More...
      </a>
    </p>
    <button
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</DocumentFragment>
`;

exports[`Rockets should render component correctly 1`] = `
<DocumentFragment>
  <div
    class="rockets-container"
  />
</DocumentFragment>
`;
